Use modern definitions for Monad and Alternative (#46)
learnyouahaskell/learnyouahaskell.github.io: markdown/source_md/a-fistful-of-monads.md

This is what the type class looks like:

```{.haskell:hs}
class Applicative m => Monad m where
    return :: a -> m a
    return = pure

    (>>=) :: m a -> (a -> m b) -> m b

    (>>) :: m a -> m b -> m b
    x >> y = x >>= \_ -> y
```

![this is you on monads](assets/images/a-fistful-of-monads/kid.png){.right width=363 height=451}

Let's start with the first line.

> It says `class Applicative => Monad m where` which means that if we want to create an instance of Monad for some time, we must have an instance of Applicative for that type.

Reviewer: ```suggestion
It says `class Applicative => Monad m where`, which means that if we want to create an instance of `Monad` for some type, we must also have an instance of `Applicative` for that type.
```

Next up, we have `>>`.
We won't pay too much attention to it for now because it comes with a default implementation and we pretty much never implement it when making `Monad` instances.

Now that we know what the `Monad` type class looks like, let's take a look at how `Maybe` is an instance of `Monad`!

```{.haskell:hs}
instance Monad Maybe where
    Nothing >>= f = Nothing
    Just x >>= f  = f x
```

> Both `return` and `(>>)` have _default implementations_, so we omit them in instances.

Reviewer: ```suggestion
Both `return` and `(>>)` have _default implementations_, so we can omit them in instances.
```

If the matching falls through all the patterns for a given function, an error is thrown and our program crashes.
On the other hand, failed pattern matching in `let` expressions results in an error being produced right away, because the mechanism of falling through patterns isn't present in `let` expressions.
When pattern matching fails in a `do` expression, the `fail` function is called.

> It's part of the `MonadFail` type class and it enables failed pattern matching to result in a failure in the context of the current monad instead of making our program crash.

Reviewer: ```suggestion
It's part of the `MonadFail` type class and it enables failed pattern matching to result in a failure in the context of the current monad, instead of making our program crash.
```

We apply `show` to `x` to turn our number into a string and then we check if the character `'7'` is part of that string.
Pretty clever.
To see how filtering in list comprehensions translates to the list monad, we have to check out the `guard` function and the `Alternative` type class.

> The `Alternative` type class is for Applicatives that can also act as monoids.

Reviewer: ```suggestion
The `Alternative` type class is for applicative functors that can also act as monoids.
```

Here's its definition:

```{.haskell:hs}
class Applicative f => Alternative f where
    empty :: f a
    (<|>) :: f a -> f a -> f a
```

`empty` is synonymous to `mempty` from the `Monoid` type class and `(<|>)` corresponds to `mappend`.
Because lists are monoids as well as monads, they can be made an instance of this type class:

```{.haskell:hs}
instance Alternative [] where
    empty = []
    (<|>) = (++)
```

> For lists `empty` represents a non-deterministic computation that has no results at all --- a failed computation.

Reviewer: ```suggestion
For lists, `empty` represents a non-deterministic computation that has no results at all --- a failed computation.
```
